### PR TITLE
Vector tiles

### PR DESCRIPTION
**Pull request by sdlime**
Hi Thomas: I have something working with WMS source and a disk cache. Just wondering if you could take a peek. I'm thinking WMS is probably the most meaningful source, but other cache types would need support. Config would look something like:

```
<format name="MVT" type="RAW">
   <extension>mvt</extension>
   <mime_type>application/x-protobuf</mime_type>
</format>

<!-- WMS returns Mapbox Vector Tiles (.mvt) -->
<source name="test_wms_mvt" type="wms">
  <http>
    <url>http://mapserver.localhost/cgi-bin/mapserv?</url>
  </http>
  <getmap>
    <params>
      <LAYERS>sna,wma,lakes,major_roads</LAYERS>
      <FORMAT>application/x-protobuf</FORMAT>
      <MAP>test.map</MAP>
    </params>
  </getmap>
</source>

<tileset name="test_mvt">
  <source>test_wms_mvt</source>
  <grid>g</grid>
  <cache>disk</cache>
  <format>MVT</format>
</tileset>
```

--Steve

### CODE REVIEW

**Comment by sdlime** [comment]
I noticed it's not setting the right mime-type on request. Still works but that's something I need to fix... --Steve

**Comment by sdlime** [comment]
Thanks for the review. I was surprised too. Once I got the tile written to the disk cache the response piece just sort of worked (just fixed mime-type). I knew robustness would be an issue. Aside from that the approach looks reasonable? Could also (maybe) enable UTFGrids through this. I thought that had been done via GSoC but I guess not. --Steve

**Comment by sdlime** [comment]
@tbonfort, I tried and this approach does work for UTFGrid too, for example:

```
<format name="UTFGRID" type="BLOB">
   <extension>json</extension>
   <mime_type>application/json</mime_type>
</format>

<!-- WMS returns UTFGrid data (.json) -->
<source name="test_wms_mvt" type="wms">
  <http>
    <url>http://mapserver.localhost/cgi-bin/mapserv?</url>
  </http>
  <getmap>
    <params>
      <LAYERS>sna,wma,lakes,major_roads</LAYERS>
      <FORMAT>application/json</FORMAT>
      <MAP>test.map</MAP>
    </params>
  </getmap>
</source>

<tileset name="test_utfgrid">
  <source>test_wms_utfgrid</source>
  <grid>g</grid>
  <cache>disk</cache>
  <format>UTFGRID</format>
</tileset>
```
Perhaps BLOB as a type is too narrow, maybe RAW instead?

Steve

**Comment by sdlime** [comment]
Testing comment (will update as I can test cache types):

1.  SQLite3: works without modification to cache_sqlite.c
2.  BDB:

**Comment by sdlime** [comment]
@tbonfort I think it's easiest just to disable WMS support for tilesets referencing raw formats. I've updated service_wms.c to do that but wanted to get your thoughts. In getTile mode with one layer it can work but that's the only case where it seems to make sense. --Steve

**Comment by sdlime** [comment]
@tbonfort, just a bump - I know you're busy.

**Comment by tbonfort** [comment]
sorry about my inactivity... ok for disabling WMS for such layers. do you want a more in depth review? cheers

**Comment by sdlime** [comment]
No worries… I wouldn’t feel comfortable merging without a thorough review. I think I hit all the integration spots but I don’t know that code like you do. Would be nice to get a new MapCache release out with this patch and whatever else is sitting in master. I suspect I owe you some tests as well – some thoughts on what those should entail would be helpful.

File: lib/source_wms.c [hunk 1]
```diff
@@ -92,7 +93,7 @@ void _mapcache_source_wms_render_map(mapcache_context *ctx, mapcache_map *map)
   mapcache_http_do_request(ctx,http,map->encoded_data,NULL,NULL);
   GC_CHECK_ERROR(ctx);
 
-  if(!mapcache_imageio_is_valid_format(ctx,map->encoded_data)) {
+  if(map->tileset->format->type != GC_RAW && !mapcache_imageio_is_valid_format(ctx,map->encoded_data)) {
```
Comment (tbonfort) [inline lib/source_wms.c]: not checking tileset->format!=NULL might be the cause of the crash in https://github.com/mapserver/mapserver/issues/5596#issuecomment-454412529_

File: lib/service_wms.c [hunk 2]
```diff
@@ -728,7 +733,7 @@ void _mapcache_service_wms_parse_request(mapcache_context *ctx, mapcache_service
       goto proxies;
     } else {
       mapcache_tileset *tileset = mapcache_configuration_get_tileset(config,str);
-      if(!tileset) {
+      if(!tileset || tileset->format->type == GC_RAW) {
```
Comment (tbonfort) [inline lib/service_wms.c]: not checking tileset->format!=NULL might be the cause of the crash in https://github.com/mapserver/mapserver/issues/5596#issuecomment-454412529_

File: lib/core.c [hunk 3]
```diff
@@ -502,6 +512,7 @@ mapcache_http_response *mapcache_core_get_map(mapcache_context *ctx, mapcache_re
   }
 
   /* compute the content-type */
+  ctx->log(ctx,MAPCACHE_DEBUG,"SDL: setting content type (2)");
```
Comment (tbonfort) [inline lib/core.c]: stray debug message to remove

**Comment by tbonfort** [comment]
LGTM once the checks for tileset->format!=NULL are added

**Comment by norac89** [comment]
After testing this pull request, it seems tiles are being cached even if they are outside the grid's restricted_extent. 
As a quick test, I have used the mapcache_seed to delete the cached tile with the same extent parameter and it deleted all tiles except those outside the restricted_extent.

**Comment by sdlime** [comment]
Can you try non-vector tiles and see if you see the same behavior with the restricted extent? I wouldn't have thought this pull request touched would have impacted that mapcache functionality but then again it's been awhile...

**Comment by norac89** [comment]
Same problem with non-vector tiles... Sorry, I've been too fast posting this here. I'll open a issue for the problem.

**Comment by sdlime** [comment]
> LGTM once the checks for tileset->format!=NULL are added (there are a couple few others I haven't ref'd directly in service_wms.c)

I think I caught all the format != NULL checks... added a little helper function.

**Pull request merged**